### PR TITLE
Feature: Add warning and collect details of URL errors in sitemap

### LINKED ISSUE #817: Enhance Bulk URL Analysis Error Handling
## Feature Request: Enhance Bulk URL Analysis Error Handling

**Description:**

Improve the bulk URL analysis tool's error handling to ensure it continues processing even when individual URLs encounter errors. Provide detailed error reporting, including the specific URL that triggered the error, to allow users to investigate further.

**Motivation:**

The current behavior of the tool, halting analysis and report generation entirely when an error occurs, hinders productivity. Users need to manually identify the problematic URLs and re-run the analysis, which is time-consuming. This enhancement will improve the tool's robustness and usability.

**User Story:**

When analyzing 1000+ URLs in bulk I want the tool to gracefully handle errors and continue processing the remaining URLs so that it can complete the full analysis and get a comprehensive report even if some URLs fail. 
I also want the tool to explicitly identify the URLs that triggered errors so that I can investigate them further if needed.

**Acceptance Criteria:**

- [ ] The tool should not stop the analysis when an error occurs for a specific URL.
- [ ] The final report should be generated even if some URLs failed.
- [ ] The report or tool output should clearly indicate which URLs caused errors.
- [ ] The error message associated with each failed URL should be included in the report or output.

**Additional Information:**

- Example error stack trace (provided):
  ```
  TypeError: Failed to set the 'text' property on 'HTMLScriptElement': This document requires 'TrustedScript' assignment.
  pptr:evaluateHandle;i.addScriptTag%20(%2Fusr%2Flocal%2F.npm-global%2Flib%2Fnode_modules%2F%40google-psat%2Fcli%2Fdist%2Fmain.js%3A3%3A961138):1:123
  ```
　　Which above won't tell which URL and it won't generate the report even other URLs has been completed the analysis.

- Related issues or discussions (if applicable)

This enhancement will significantly improve the user experience by making the bulk URL analysis tool more resilient and informative.

### LINKED ISSUE #818: PSAT CLI run in loop when doing analysis of sitemaps or CSV files
**Describe the bug**
We encountered an issue where the PSAT CLI enters a never-ending loop when processing specific sitemaps or CSV files. This might be due to the CLI failing to detect redirections or handling different redirection HTTP statuses correctly.

My initial findings suggest this may be because of redirection used by websites, but even if they are redirection it should not run in endless loop.

**To Reproduce**
- Open terminal 
- Sitemap
  - Use the command `npm run cli -- -s https://css-tricks.com/post-sitemap10.xml -n 50`
- CSV file: (attached below)
  - `npm run cli -- -f /Users/milind/Downloads/http-status.csv` 
  - Observe if the CLI gets stuck in a loop and doesn't proceed to other URLs.

**Expected behavior**
The PSAT CLI should process the provided sitemap and CSV file, handle any redirects encountered, and continue processing other URLs in the queue.

**Screenshots**

https://github.com/user-attachments/assets/47b27024-a948-4a24-bfc5-1f1b41b61f12

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser : Chrome
 - Version : 0.10.1

**Additional context**
CSV File 
[http-status.csv](https://github.com/user-attachments/files/16729959/http-status.csv)

### PR DESCRIPTION
**Pull request by amovar18**
## Description
This PR aims to collect errors on URL which occur while analysing the sitemap.

## Relevant Technical Choices
- The CLI catches critical errors when sitemap analysis is running.
- These errors are accumulated, processed and then these URLs are displayed on the CLI-Dashboard.

## Testing Instructions
- Clone this branch.
- In the terminal, run `npm i` and run `npm run build:cli:dashboard`.
- Use a CSV or a sitemap where there are URLs with errors.
- After analysis of the sitemap, open the dashboard, you'll notice those URLs listed under URLs with errors. 

**Example:**
CSV File: [psat-csv-test.csv](https://github.com/user-attachments/files/16903761/psat-csv-test.csv)
Run `npm run cli -- -f ~/Downloads/psat-csv-test.csv`

## Screenshot/Screencast

---

## Checklist

- [x] I have thoroughly tested this code to the best of my abilities.
- [x] I have reviewed the code myself before requesting a review.
- [x] This code is covered by unit tests to verify that it works as intended.
- [x] The QA of this PR is done by a member of the QA team (to be checked by QA).

Fixes #818

### CODE REVIEW

**Comment by pavanpatil1** [comment]
- [ ] The icons for the URLs with issues are not displaying correctly.

Screenshot 2024-09-03 at 2 40 31 PM

---

- [ ] If a URL from the CSV or sitemap fails, the CLI results currently do not display a message. We should add a message that says, 'Some URLs encountered issues and were not executed properly. You can find a details on dashbord'  when any URLs fail during execution

Screenshot 2024-09-03 at 3 04 11 PM

File: packages/cli/src/utils/getSiteReport.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * @param processedData The cookie data along with library detection information with erroredOutUrls.
+ * @param technologyAnalysisData The technology analysis data from CLI.
+ * @returns {object} The object which will be used to send the report to the cli dashboard.
+ */
+function getSiteReport(
+  urls: string[],
+  processedData: any,
+  technologyAnalysisData: any
+) {
+  return urls.map((_url, ind) => {
```
Comment (mohdsayed) [inline packages/cli/src/utils/getSiteReport.ts:44]:
```suggestion
  return urls.map((url, ind) => {
```

File: packages/cli/src/utils/getSiteReport.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    const hasTimeOutError = (
+      processedData[ind].erroredOutUrls[_url] as SingleURLError[]
```
Comment (mohdsayed) [inline packages/cli/src/utils/getSiteReport.ts:46]:
```suggestion
    const errorOutUrls = processedData[ind].erroredOutUrls[url];
    const hasTimeOutError = (
      errorOutUrls as SingleURLError[]
```

File: packages/report/src/dashboard/components/urlsWithIssues/index.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+          <span className="w-full flex justify-center">{info}</span>
+        ),
+      },
+    ],
+    []
+  );
+
+  const filters = useMemo<TableFilter>(() => ({}), []);
+
+  const searchKeys = useMemo<string[]>(() => ['url'], []);
```
Comment (mohdsayed) [inline packages/report/src/dashboard/components/urlsWithIssues/index.tsx:66]: What's the purpose of using `useMemo` here?

Comment (amovar18) [reply packages/report/src/dashboard/components/urlsWithIssues/index.tsx:66]: It was being used in other places as well. Will change that.

File: packages/report/src/dashboard/components/urlsWithIssues/index.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+  const tablePersistentSettingsKey = useMemo<string>(() => {
+    return 'urlsWithIssues';
+  }, []);
```
Comment (mohdsayed) [inline packages/report/src/dashboard/components/urlsWithIssues/index.tsx:70]: `useMemo` ?

File: packages/cli/src/index.ts [hunk 5]
```diff
@@ -271,23 +279,11 @@ program.parse();
...
-      technologyData: technologyAnalysisData ? technologyAnalysisData[ind] : [],
-      cookieData: cookieAnalysisAndFetchedResourceData[ind].cookieData,
-      libraryMatches: detectedMatchingSignatures ?? [],
-    } as unknown as CompleteJson;
-  });
+  const result = getSiteReport(
+    urlsToProcess,
+    cookieAnalysisAndFetchedResourceData,
+    technologyAnalysisData
+  );
```
Comment (mohdsayed) [inline packages/cli/src/index.ts:286]: 👍

**Review by mohdsayed — APPROVED** [review]
Looks good. Just a few comments to address.

File: packages/report/src/dashboard/components/urlsWithIssues/index.tsx (hunk 4 above)
Comment (amovar18) [reply packages/report/src/dashboard/components/urlsWithIssues/index.tsx]: Resolved in latest commit.

File: packages/cli/src/utils/getSiteReport.ts (hunk 2 above)
Comment (amovar18) [reply packages/cli/src/utils/getSiteReport.ts]: Resolved in latest commit.

File: packages/cli/src/utils/getSiteReport.ts (hunk 1 above)
Comment (amovar18) [reply packages/cli/src/utils/getSiteReport.ts]: Resolved in latest commit.

**Pull request merged**